experimental/report/renderer.go

```diff
@@ -0,0 +1,893 @@
+	// for the same line into rendered chunks
+	parts := partition(w.underlines, func(a, b *underline) bool { return a.line != b.line })
+	parts(func(_ int, part []underline) bool {
+		cur := &info[part[0].line-w.start]
+		cur.shouldEmit = true
+
+		// Arrange for a "sidebar prefix" for this line. This is determined by any sidebars that are
+		// active on this line, even if they end on it.
+		sidebar := renderSidebar(sidebarLen, -1, -1, c, cur.sidebar)
+
+		// Lay out the physical underlines in reverse order. This will cause longer lines to be
+		// laid out first, which will be overwritten by shorter ones.
+		//
+		// We use a slice instead of a strings.Builder so we can overwrite parts
+		// as we render different "layers".
+		var buf []byte
+		for i := len(part) - 1; i >= 0; i-- {
+			element := part[i]
+			if len(buf) < element.end {
+				newBuf := make([]byte, element.end)
+				copy(newBuf, buf)
+				buf = newBuf
+			}
+
+			// Note that start/end are 1-indexed.
+			for j := element.start - 1; j < element.end-1; j++ {
+				buf[j] = byte(element.level)
+			}
+		}
+
+		// Now, convert the buffer into a proper string.
+		var out strings.Builder
+		parts := partition(buf, func(a, b *byte) bool { return *a != *b })
+		parts(func(_ int, line []byte) bool {
+			level := Level(line[0])
+			if line[0] == 0 {
+				out.WriteString(c.reset)
+			} else {
+				out.WriteString(c.BoldForLevel(level))
+			}
+			for range line {
+				switch level {
+				case 0:
+					out.WriteByte(' ')
+				case note:
+					out.WriteByte('-')
+				default:
+					out.WriteByte('^')
+				}
+			}
+			return true
+		})
+
+		// Next we need to find the message that goes inline with the underlines. This will be
+		// the message belonging to the rightmost underline.
+		var rightmost *underline
+		for i := range part {
+			ul := &part[i]
+			if rightmost == nil || ul.end > rightmost.end {
+				rightmost = ul
+			}
+		}
+		underlines := strings.TrimRight(out.String(), " ")
+		cur.underlines = []string{sidebar + underlines + " " + c.BoldForLevel(rightmost.level) + rightmost.message}
+
+		// Now, do all the other messages, one per line. For each message, we also
+		// need to draw pipes (|) above each one to connect it to its underline.
+		//
+		// This is slightly complicated, because there are two layers: the pipes, and
+		// whatever message goes on the pipes.
+		var rest []*underline
+		for i := range part {
+			ul := &part[i]
+			if ul == rightmost || ul.message == "" {
+				continue
+			}
+			rest = append(rest, ul)
+		}
+
+		for idx := range rest {
+			buf = buf[:0] // Clear the temp buffer.
+
+			// First, lay out the pipes. Note that rest is not necessarily
+			// ordered from right to left, so we need to sort the pipes first.
+			// To deal with this, we make a copy of rest[idx:], sort it appropriately,
+			// and then lay things out.
+			//
+			// This is quadratic, but no one is going to put more than like. five snippets
+			// in a line, so it's fine.
```

Comment:
I don't think there's a good way to avoid it being quadratic. Generally, I'm not sure it's worth putting in tons of effort for cases that will generate really unreadable diagnostics. You won't notice a slowdown until you have dozens of messages in a single line, at which point the diagnostic is unreadable and you're just hitting yourself.

Reply:
My point is that this case will never happen in our compiler because doing so would violate the style guide I wrote. It's not a thing that can happen "accidentally": the number of messages per diagnostic should never be linearly dependent on e.g. user input.

Trying to make this work means more cases and more tests for no benefit to us.